pkg/operator/status.go

```diff
@@ -66,182 +92,225 @@ func (r *StatusReporter) GetOrCreateClusterOperator() (*configv1.ClusterOperator
 		},
 	}
 
-	existing, err := r.client.ConfigV1().ClusterOperators().
-		Get(OperatorName, metav1.GetOptions{})
+	existing, err := r.GetClusterOperator()
 
 	if errors.IsNotFound(err) {
-		return r.client.ConfigV1().ClusterOperators().Create(clusterOperator)
+		return r.configClient.ConfigV1().ClusterOperators().Create(clusterOperator)
 	}
 
 	return existing, err
 }
 
-// ApplyConditions applies the given conditions to the ClusterOperator
-// resource's status.
-func (r *StatusReporter) ApplyConditions(conds []configv1.ClusterOperatorStatusCondition, reachedLevel bool) error {
-	status := configv1.ClusterOperatorStatus{
-		RelatedObjects: r.relatedObjects,
+// ApplyStatus applies the given ClusterOperator status to the operator's
+// ClusterOperator object if necessary.  The currently configured RelatedObjects
+// are automatically set on the status.  If no ClusterOperator objects exists,
+// one is created.
+func (r *StatusReporter) ApplyStatus(status configv1.ClusterOperatorStatus) error {
+	var modified bool
+
+	co, err := r.GetOrCreateClusterOperator()
+	if err != nil {
+		return err
 	}
 
-	for _, c := range conds {
-		cvorm.SetOperatorStatusCondition(&status.Conditions, c)
+	// Set the currently configured related objects.
+	status.RelatedObjects = r.config.RelatedObjects
+
+	// If no versions were set explicitly, continue reporting previous versions.
+	if status.Versions == nil {
+		status.Versions = co.Status.Versions
 	}
 
-	if reachedLevel {
-		status.Versions = []configv1.OperandVersion{
-			{
-				Name:    "operator",
-				Version: r.releaseVersion,
-			},
+	// Update LastTransitionTime for all conditions if necessary.
+	for i := range status.Conditions {
+		condType := status.Conditions[i].Type
+		timestamp := metav1.NewTime(time.Now())
+
```

Comment:
we should standardise on https://github.com/openshift/library-go/blob/master/pkg/config/clusteroperator/v1helpers/status.go